File: examples/language-java/src/main/java/example/errors/LiveLocalAfterMessage.java
package example.errors;

/**
 * The <em>live-local</em> counterpart to {@link SlowMessage}: a method that guards its argument with a
 * {@code throw new IllegalArgumentException(<dynamic message>)} AND, on the normal path, carries a value
 * through a LOCAL that it returns and the caller goes on to use. The dynamic message makes
 * exception-message elision rewrite this method; the returned local is genuinely LIVE.
 *
 * <p>This is the exact shape of {@code Bmc.anyString(min, max)} (a {@code throw new IAE("...got " + n)}
 * argument check, then {@code String s = ...; return s;}). It pins the soundness boundary that eliding a
 * message must leave the rest of the method - its live locals and their type/non-null tracking - untouched.
 * The value carried is a SYMBOLIC string (the caller passes {@code Bmc.anyAsciiString(...)}): its non-null
 * guarantee comes from the engine's nondet modeling, which reads the method's LocalVariableTable. A rewrite
 * that drops a modified method's LocalVariableTable wholesale makes the engine lose that non-null tracking,
 * so the caller reads the returned string back {@code null} and a correct proof FALSE-REFUTES with a
 * NullPointerException.
 */
public final class LiveLocalAfterMessage {

    private LiveLocalAfterMessage() {
    }

    /**
     * Rejects a {@code null} {@code s} with a dynamic message (the elided construction), otherwise carries
     * {@code s} through a local and returns it. The returned value is LIVE: the caller reads its
     * {@code length()}. Eliding the message must not disturb it.
     */
    public static String checkedLabel(String s, int tag) {
        if (s == null) {
            throw new IllegalArgumentException("label must be non-null, tag=" + tag);
        }
        String label = s;
        return label;
    }
}

File: examples/language-kotlin/src/main/kotlin/example/exceptions/LiveLocalAfterMessage.kt
package example.exceptions

/**
 * The live-local counterpart of an elided exception message, in idiomatic Kotlin: a function that guards
 * its argument with a `throw IllegalArgumentException(<string template>)` and then, on the normal path,
 * carries a value through a LOCAL it returns for the caller to use.
 *
 * The string template compiles to a `makeConcatWithConstants` invokedynamic, so exception-message elision
 * rewrites this function; the returned `label` is genuinely LIVE. It pins that eliding the message leaves
 * the rest of the function - its live locals and their non-null tracking - untouched. The value carried is
 * a SYMBOLIC string (the caller passes `Bmc.anyAsciiString(...)`), whose non-null guarantee comes from the
 * engine's nondet modeling via the LocalVariableTable. A rewrite that drops a modified method's
 * LocalVariableTable wholesale made the returned `String` read back null, so a caller that reads
 * `label.length` FALSE-REFUTED with a NullPointerException over correct code.
 */
fun checkedLabel(s: String?, tag: Int): String {
    if (s == null) throw IllegalArgumentException("label must be non-null, tag=$tag")
    val label = s
    return label
}
